### DISCUSSION TITLE
Dispatching a livewire event from a filament action when submit

### DISCUSSION THREAD

**Discussion by mahmoud-757** [filamentphp/filament#11675] [Help]
### Package

Panel builder

### Package Version

v3.2.35

### How can we help you?

I am trying to dispatch a refresh event when a modal action is submitted
the event is dispatched from other functions inside the component except from actions
here is the code
```
protected $listeners = ['refreshTable' => 'refresh'];

public function chooseAssessment(): Action
    {
    return Action::make('chooseAssessment')
            ->label(__('Choose Assessment'))
            ->color('info')
            ->form([
             // some form components
            ])
            // ->extraAttributes(['wire:click' => '$dispatch(\'refreshTable\')'])
            ->modalWidth('md')
            ->action(function (array $data, \Livewire\Component $livewire) {
                // doing some stuff with $data
                
                // $livewire->dispatch('refreshTable')->component(ThisClassName::class);
                // $this->dispatch('refreshTable');
                // $livewire->dispatch('refreshTable');
            })
            // ->after(function (\Livewire\Component $livewire) {
                // $this->dispatch('refreshTable');
                // $livewire->dispatch('refreshTable');
                // $livewire->dispatch('refreshTable')->component(ThisClassName::class);
            // })
            ;
}
```
            
all the commented code is what I have tried but with no luck
and I also tried `#[On('refreshTable')]` on the component 

this is inside a filament resource custom page created with this command
`php artisan make:filament-page SortUsers --resource=UserResource --type=custom`

I am creating a new discussion as I've tried all the recommended solutions found on this topic 
Thanks in advance!

Edit: Added a video explaining the problem in the comments below

**Comment by leandrocfe** [comment]
Where are you using this action? Are you rendering this action [like this](https://filamentphp.com/docs/3.x/actions/modals#modal-forms)?

**Reply by mahmoud-757** [reply]
Thanks for the quick reply.
I am using it in a resource custom page `app/Filament/Resources/ResourceName/Pages/ClassName.php` and displaying it in the view like this
 ```                        
<div>
 {{ $this->chooseAssessment() }}
</div>
```
is that what you mean?
And yes I am using it like the docs and everything else is working except dispatching events from inside the action

**Reply by mahmoud-757** [reply]
https://github.com/filamentphp/filament/assets/147031958/5ce92087-bcc1-4175-959a-ffead0faa6cc

here is a video showcasing my problem
everything is reactive and is re-rendered correctly some automatically (visibility button) and some via events (stop assessment button) except when I click choose assessment even creating a new one refreshes the table via dispatch event and it works fine so I don't know what to do xD

**Reply by mahmoud-757** [reply]
Was this what you were asking for? @leandrocfe

**Reply by leandrocfe** [reply]
try `dispatch`
https://github.com/filamentphp/filament/discussions/11674#discussioncomment-8666907

**Reply by mahmoud-757** [reply]
First of all thanks for taking the time to help.

I did it like this
```
            ->action(function (array $data, \Livewire\Component $livewire) {
                // doing some stuff with $data

                Action::make('refreshTable')
                    ->dispatchSelf('refreshTable');
            })
```
and like this
```
protected $listeners = ['refreshTable' => 'refresh'];

public function chooseAssessment(): Action
    {
    return Action::make('chooseAssessment')
            ->label(__('Choose Assessment'))
            ->color('info')
            ->form([
             // some form components
            ])
            ->modalWidth('md')
            ->action(function (array $data, \Livewire\Component $livewire) {
                // doing some stuff with $data
            })
            ->dispatchSelf('refreshTable'); // added the dispatch here also tried dispatch() but no luck
}
```
both have the same behavior which is the button not showing the modal when clicked

https://github.com/filamentphp/filament/assets/147031958/a7b0ec7a-17ba-486b-8e96-6784c3c4a99e

**Comment by mahmoud-757** [comment]
Any updates?

**Comment by BatthewZ** [comment]
I might be having a similar issue, where I've got a dispatch in an EditModel page's afterSave() method, that is listened to successfully by a relationsManager. However, trying to get an action in the relationsManager to dispatch an event and have it be successfully heard by the ModelResource is proving impossible and I have no idea why... before `dispatch()` there was `emit()` and `emitUp()`. I wonder if it's something to do with parents and children, or sibling components and how that wires together behind the scenes...?

**Comment by acharseth** [comment]
@mahmoud-757 , where have you placed the `chooseAssessment()`.
I have a similar problem where I try this:

```
use Filament\Notifications\Actions\Action as NotificationAction;
...
NotificationAction::make('confirm')
->label('Confirm')
->button()
->color('danger')                                          
->dispatch('generateNewCodeEvent', [$get('code'), $get('code_length')])                                                                                  
->close(),  
```

I have put `generateNewCode()` in a Livewire component class:      

```
use Livewire\Component;

class CodeGen extends Component
{
    public function render()
    {
        return view('livewire.code-gen');
    }
    
    #[On('generateNewCodeEvent')]
    public function generateNewCode($currentCode, $codeLen)
    {                                        
  ...
  ```
  `generateNewCode()` never gets called, so I wonder if I need to register/activate CodeGen somewhere?
